pallets/deitos/src/lib.rs

```diff
@@ -91,20 +95,38 @@ pub mod pallet {
         Transfer,
     }
 
+    #[pallet::genesis_config]
+    #[derive(frame_support::DefaultNoBound)]
+    pub struct GenesisConfig<T: Config> {
+        /// Genesis Initial IP Bounding
+        pub initial_ip_bounding: BalanceOf<T>,
+    }
+
+    #[pallet::genesis_build]
+    impl<T: Config> BuildGenesisConfig for GenesisConfig<T> {
+        fn build(&self) {
+            IPBoundingAmount::<T>::put(&self.initial_ip_bounding);
+        }
+    }
+
+    #[pallet::storage]
+    #[pallet::getter(fn ip_bounding_amount)]
+    pub type IPBoundingAmount<T: Config> =
+        StorageValue<_, BalanceOf<T>, ResultQuery<Error<T>::NonExistentStorageValue>>;
+
     #[pallet::storage]
     #[pallet::getter(fn get_ip)]
     pub type InfrastructureProvider<T: Config> =
         StorageMap<_, Blake2_128Concat, T::AccountId, InfraProviderDetails<T>>;
 
     #[pallet::storage]
-    #[pallet::getter(fn get_agreements)]
-    pub type Agreements<T: Config> = StorageNMap<
+    #[pallet::getter(fn agreements)]
+    pub(super) type Agreements<T: Config> = StorageDoubleMap<
```

Comment:
I assume that you are referring to  have a map like `consumer/ip => agreement_id`  other wise we would be deplicating data on the storage which we should avoid of course.

I wonder that if we might avoid having a separate storage item for querying purpose by relaying on this:

https://paritytech.github.io/polkadot-sdk/master/src/frame_support/storage/mod.rs.html#714-716

```
	/// Does any value under the first key `k1` (explicitly) exist in storage?
	/// Might have unexpected behaviour with empty keys, e.g. `[]`.
	fn contains_prefix<KArg1>(k1: KArg1) -> bool
	where
		KArg1: EncodeLike<K1>;
```

Reply:
If we mix it in a single map of `consumer/ip => agreement_id` then the agreements will be mixed up for wallet that have both roles IP and customer. Also you already have `ip => agreement_id` map (InfrastructureProvider), so we just need a single storage that maps customers to agreements.

Reply:
ah no... i didn't mean to have one storage item for that mapping. I understood that you proposal was to have two different StorageValueMap like:

`ip -> agreement_id`
`consumer -> agreement_id`

We might ommit the first one, but i see the need for the cosumers one.

Reply:
ah ok , i think i'm getting your idea ... you are propsing instead of having a StorageDoubleMap you propose:

`ip -> AgreementDetails<T>`
`consumer -> AgreementId`

Reply:
You already have `InfrastructureProvider` which is map `ip => InfraProviderDetails`. `InfraProviderDetails` stores `active_agreements` which is a vector of `AgreementId`. Where is a map that maps `AgreementId` to `AgreementDetails`?

So I think `InfrastructureProvider` is fine, but `Agreements` should be `AgreementId => AgreementDetails`. Next you need to consumers know there current agreements too. So we need 1 extra map of `customer => AgreementId`.

Reply:
discussed privately and agreed. To be worked in the next PR.